codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketUtil.java

```diff
@@ -96,10 +96,18 @@ static String base64(byte[] data) {
         if (PlatformDependent.javaVersion() >= 8) {
             return java.util.Base64.getEncoder().encodeToString(data);
         }
+        String encodedString;
         ByteBuf encodedData = Unpooled.wrappedBuffer(data);
-        ByteBuf encoded = Base64.encode(encodedData);
-        String encodedString = encoded.toString(CharsetUtil.UTF_8);
-        encoded.release();
+        try {
+            ByteBuf encoded = Base64.encode(encodedData);
+            try {
+                encodedString = encoded.toString(CharsetUtil.UTF_8);
```

Comment:
@seedeed all of this should never throw. If it throws we have a bug in our implementation. What I am missing ?

Reply:
Not our implementation bug, I think... But the `Base64.encode` may throw an `OutOfDirectMemoryError` if the jvm argument `-XX:MaxDirectMemorySize`'s value is too small. For instance:

```java
Exception in thread "main" io.netty.util.internal.OutOfDirectMemoryError: failed to allocate 28 byte(s) of direct memory (used: 1048576, max: 1048576)
	at io.netty.util.internal.PlatformDependent.incrementMemoryCounter(PlatformDependent.java:742)
	at io.netty.util.internal.PlatformDependent.allocateDirectNoCleaner(PlatformDependent.java:697)
	at io.netty.buffer.UnpooledUnsafeNoCleanerDirectByteBuf.allocateDirect(UnpooledUnsafeNoCleanerDirectByteBuf.java:30)
	at io.netty.buffer.UnpooledByteBufAllocator$InstrumentedUnpooledUnsafeNoCleanerDirectByteBuf.allocateDirect(UnpooledByteBufAllocator.java:186)
	at io.netty.buffer.UnpooledDirectByteBuf.<init>(UnpooledDirectByteBuf.java:64)
	at io.netty.buffer.UnpooledUnsafeDirectByteBuf.<init>(UnpooledUnsafeDirectByteBuf.java:41)
	at io.netty.buffer.UnpooledUnsafeNoCleanerDirectByteBuf.<init>(UnpooledUnsafeNoCleanerDirectByteBuf.java:25)
	at io.netty.buffer.UnpooledByteBufAllocator$InstrumentedUnpooledUnsafeNoCleanerDirectByteBuf.<init>(UnpooledByteBufAllocator.java:181)
	at io.netty.buffer.UnpooledByteBufAllocator.newDirectBuffer(UnpooledByteBufAllocator.java:91)
	at io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:187)
	at io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:178)
	at io.netty.buffer.AbstractByteBufAllocator.buffer(AbstractByteBufAllocator.java:115)
	at io.netty.handler.codec.base64.Base64.encode(Base64.java:108)
	at io.netty.handler.codec.base64.Base64.encode(Base64.java:100)
	at io.netty.handler.codec.base64.Base64.encode(Base64.java:80)
	at io.netty.handler.codec.base64.Base64.encode(Base64.java:70)
	at io.netty.handler.codec.base64.Base64.encode(Base64.java:66)
	at io.netty.handler.codec.http.websocketx.WebSocketUtil.base64(WebSocketUtil.java:100)
```

Reply:
@seedeed good point...